src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java

```diff
@@ -1284,6 +1384,107 @@ public HttpResponse doFinishLogin(StaplerRequest request) throws IOException {
         return currentSession.finishLogin(request, buildAuthorizationCodeFlow());
     }
 
+    /**
+     * Handles Token Expiration.
+     *
+     * @throws IOException a low level exception
+     */
+    public void handleTokenExpiration(HttpServletRequest httpServletRequest) throws IOException {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        if (authentication == null || !(authentication.getCredentials() instanceof OicCredentials)) {
+            return;
+        }
+
+        OicCredentials credentials = (OicCredentials) authentication.getCredentials();
+
+        if (credentials.isExpired()) {
+            if (isUseRefreshTokens() && credentials.getRefreshToken() != null) {
+                refreshExpiredToken(credentials, httpServletRequest);
+            } else if (isStrictTokenExpirationCheck()) {
+                throw HttpResponses.errorWithoutStack(401, "Token expired");
```

Comment:
To avoid repeating myself, all the comments reference outdated code except this one.

Sadly, I'm conflicted what to do when logging in via Browser or via an API Key.